Added timezones

thomastay opened the pull request:
Added the ability to display two clocks on the date module and the ability to set the timezone on those two clocks. This solves #596 and also addresses the problem of #788 in that the fix does not use any external modules, instead relying on setting the TZ environment variable. 

Currently the options for customization are such that users can set the setting **time-TZ** and **time-2-TZ** on their polybar config file to set the timezones of two separate clocks. The options date-2 and time-2 can be added to the label to give the ability to display a dual clock. date-2 is set using the time-2-TZ variable and is not independent of time-2. 

Here's an example using my polybar config file:
```
type = internal/date
interval = 1

date =  %a, %b %d
date-alt =  %Y-%m-%d 

time = "%H:%M"
time-TZ = "Asia/Singapore"
time-2-TZ = "America/Detroit"
time-alt = "%H:%M:%S"

label = %date%   %time% | %date-2%   %time-2%
```
![My config](https://user-images.githubusercontent.com/7616132/43384560-fec32e9e-9410-11e8-8def-d224d1974a20.png)

Minor bugs: currently, setting the option time-2-TZ but _not_ time-TZ results in both clocks having the secondary time zone. I don't know how to fix this bug at the moment.

Improvements: I intend to work on a version of this module that allows users to have different timezone when they click on the module, instead of just changing the format.

Fixes #596 
Fixes #1293

patrick96 commented:
What is the benefit of dual clocks in the same module over having two separate date modules?

thomastay commented:
@patrick96 I suppose you're right. I originally designed the dual clock system with the intention of having a click-to-change timezone module, that's why I implemented the option for dual clock. However that proved difficult without completely rewriting the module so that's why I added the option to display two clocks at once.

thomastay commented:
hmm I tried having two modules with one timezone each, but I ran into a problem that having two date modules means that the click handler only applies to first of the modules defined. This means that on clicking either of the modules, only the first module defined changes to 

Is this a known bug? I've looked through the code and I'm not sure why the

patrick96 commented:
> Is this a known bug?

Yes, this is the same issue as #1172. If you have multiple modules of the same kind that have internal click events, they are all caught and processed by the first module and don't reach the others.

thomastay commented:
Ok! Have done so.

thomastay commented:
@patrick96 are there any more changes that need to be made to the PR before it can be accepted?

patrick96 commented:
Probably. I haven't had the time to properly review this.

patrick96 commented on src/modules/date.cpp:
```diff
@@ -41,6 +42,14 @@ namespace modules {
+    const char* timezone_format_original = getenv("TZ");
+
+    string timezone_format = m_time_timezone;
+    if (timezone_format != ""){ 
```
Just a style thing: Please use a space after the if parenthesis

NBonaparte commented:
We might need to use a mutex to prevent race conditions between multiple modules.

thomastay commented:
@patrick96 ok i will look into it. Using a mutex like @NBonaparte mentioned is probably the best solution.

thomastay commented:
@patrick96 I've thought about this and here's a solution I've come up with to prevent the race condition (it is indeed a race condition, I checked). Wanted to hear your input on this as I'm not an expert on multithreading.

Have a central mutex held by the bar's _controller_. When a date module wants to access / set an environment variable, it needs to use the mutex held by the bar's controller. This ensures that only one module is accessing the "TZ" environment variable at once. 

Right now the modules cannot directly call this central mutex since they don't inherit a pointer to the controller. I was thinking of modifying the start function to include a pointer to the controller that started it, but that might be unwieldy. Any ideas / opinions?

patrick96 commented:
Yeah right now we don't really have central place to store such things where everyone has access to. I think storing the mutex in `bar_settings` in `types.hpp` for now should suffice, every module already has access to that. At some point we should probably rethink how things are coupled in general and maybe use a service locator for such things.

patrick96 commented:
@thomastay are you still working on this?

thomastay commented:
sorry, I haven't had time to work on it. Not working on it for now.

patrick96 commented:
No worries. Maybe someone else will pick this up in the future :)
